File: 0130-surrounded-regions/0130-surrounded-regions.cpp
class Solution {
public:
    void dfs(int row, int col, vector<vector<char>>& mat,vector<vector<int>>& visited) {

        visited[row][col] = 1;

        int n = mat.size();
        int m = mat[0].size();

        int drow[] = {0, 0, 1, -1};
        int dcol[] = {1, -1, 0, 0};

        for (int i = 0; i < 4; i++) {
            int nrow = row + drow[i];
            int ncol = col + dcol[i];

            if (nrow >= 0 && ncol >= 0 && ncol < m && nrow < n &&
                mat[nrow][ncol] == 'O' && !visited[nrow][ncol]) {
                dfs(nrow, ncol, mat, visited);
            }
        }
    }
    void solve(vector<vector<char>>& mat) {
        int n = mat.size();
        int m = mat[0].size();

        vector<vector<int>> visited(n, vector<int>(m, 0));
        //vector<vector<char>> sur = mat;

        for (int i = 0; i < m; i++) {
            if (mat[0][i] == 'O' && !visited[0][i]) {
                dfs(0, i, mat, visited);
            }

            if (mat[n - 1][i] == 'O' && !visited[n - 1][i]) {
                dfs(n - 1, i, mat, visited);
            }
        }

        for (int i = 0; i < n; i++) {
            if (mat[i][0] == 'O' && !visited[i][0]) {
                dfs(i, 0, mat, visited);
            }

            if (mat[i][m - 1] == 'O' && !visited[i][m - 1]) {
                dfs(i, m - 1, mat, visited);
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (mat[i][j] == 'O' && !visited[i][j]) {
                    //sur[i][j] = 'X';
                    mat[i][j] = 'X';
                }
            }
        }

    }
};

File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    public void dfs(int row, int col, char[][] board, int[][] visited) {
        visited[row][col] = 1;

        int n = board.length;
        int m = board[0].length;

        int[] drow = {1,-1,0,0};
        int[] dcol = {0,0,1,-1};

        for(int i = 0;i<4;i++){
            int nrow = row + drow[i];
            int ncol = col + dcol[i];


            if(nrow >= 0 && ncol >= 0 && nrow < n && ncol < m && board[nrow][ncol] == 'O' && visited[nrow][ncol] == 0){
                dfs(nrow,ncol,board,visited);
            }
        }
    }
    public void solve(char[][] board) {
        int n = board.length;
        int m = board[0].length;

        int[][] visited = new int[n][m];

         for(int i = 0;i<m;i++){
            if(board[0][i] == 'O' && visited[0][i] == 0){
                dfs(0,i,board,visited);
            }

            if(board[n-1][i] == 'O' && visited[n-1][i] == 0){
                dfs(n-1,i,board,visited);
            }
        }

        for(int i = 0;i<n;i++){
            if(board[i][0] == 'O' && visited[i][0] == 0){
                dfs(i,0,board,visited);
            }

            if(board[i][m-1] == 'O' && visited[i][m-1] == 0){
                dfs(i,m-1,board,visited);
            }
        }


        for(int i = 0;i<n;i++){
            for(int j = 0;j<m;j++){
                if(board[i][j] == 'O' && visited[i][j] == 0){
                    board[i][j] = 'X';
                }
            }
        }
        
    }
}